namer/consul/src/main/scala/io/buoyant/namer/consul/LookupCache.scala

```diff
@@ -33,60 +38,85 @@ private[consul] class LookupCache(
-          case ((dcName, domainOption)) =>
-            SvcAddr(
-              consulApi,
-              LookupCache.DefaultBackoffs,
-              dcName,
-              key,
-              domainOption,
-              consistency = consistency,
-              preferServiceAddress = preferServiceAddress,
-              weights,
-              serviceStats
-            )
-        }
+  /*
+   * use a shared mutex only on write, the ConcurrentHM impl should guarantee
+   * happens-before semantic of reads wrt to updates (as per docs)
+   */
+  private[this] val lookupStatusMu = new {}
+  private[this] val lookupStatus = new ConcurrentHashMap[Path, InstrumentedAddr]()
 
-        val observation = Var.async[Activity.State[NameTree[Name.Bound]]](Activity.Pending) { observationState =>
-          val closableFuture = addrFuture.transform {
-            case Return(addr) =>
-              val observationClosable = addr.changes.respond {
-                case Addr.Neg => observationState.update(Activity.Ok(NameTree.Neg))
-                case Addr.Pending => observationState.update(Activity.Pending)
-                case Addr.Failed(why) => observationState.update(Activity.Failed(why))
-                case Addr.Bound(_, _) => observationState.update(Activity.Ok(NameTree.Leaf(Name.Bound(addr, id, residual))))
-              }
-              Future.value(observationClosable)
-            case Throw(cause) =>
-              // We probably failed to fetch agent config. This is critical.
-              // TODO: if this has happened only restart can restore namerd to working state. Throw exception instead?
-              observationState.update(Activity.Failed(cause))
-              Future.value(Closable.nop)
-          }
+  private[consul] def status: Map[Path, InstrumentedAddr] = lookupStatus.asScala.toMap
 
-          Closable.make { deadline =>
-            closableFuture.flatMap(_.close(deadline))
+  private[this] val lookupAddress: (Path, String, SvcKey) => Future[InstrumentedVar[Addr]] =
+    (raw, dc, key) => {
+      val pollState = SvcAddr.mkConsulPollState
+      val addrFuture: Future[InstrumentedVar[Addr]] = resolveDc(dc).join(domain).map {
+        case ((dcName, domainOption)) =>
+          SvcAddr(
+            consulApi,
+            LookupCache.DefaultBackoffs,
+            dcName,
+            key,
+            domainOption,
+            consistency = consistency,
+            preferServiceAddress = preferServiceAddress,
+            weights,
+            serviceStats,
+            pollState
+          )
+      }
+
+      //if the lookup succeeded we want to cache the result, double checking for race conditions
+      addrFuture.map {
+        addr =>
+          lookupStatusMu.synchronized {
+            if (!lookupStatus.containsKey(raw)) {
+              lookupStatus.put(raw, InstrumentedAddr(addr, pollState))
+              cachedCounter.incr()
+            }
+            lookupStatus.get(raw).addr
```

Comment:
I think this is always going to be the same value as was just put into cache so this can simply be `addr`.  In fact, since this is the original value of the future anyway, this whole `map` block can be replaced by an `onSuccess` block.

Reply:
will do